File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        
        def subCheck(p, q):
            if not p and not q:
                return True
            
            if not p or not q or p.val != q.val:
                return False
            
            return subCheck(p.left, q.left) and subCheck(p.right, q.right)
        
        def subTraversal(root):
            if not root:
                return False
            
            if subCheck(root, subRoot):
                return True
            
            return subTraversal(root.left) or subTraversal(root.right)
        
        return subTraversal(root)

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func maxPathSum(root *TreeNode) int {
	summ := math.MinInt32
    var check func(root *TreeNode) int

	check = func(node *TreeNode) int  {
		if node == nil {
			return 0
		}

		left := check(node.Left)
		right := check(node.Right) 
		leftMax := max(left, 0)
		rightMax := max(right, 0)

		currSum := node.Val + leftMax + rightMax
		summ = max(summ, currSum)
    
		return node.Val + max(leftMax, rightMax) 
	}

    check(root)
	return summ
}
